File: codewars.com/javascript/5kyu/calculating-functions.js
// https://www.codewars.com/kata/525f3eda17c7cd9f9e000b39/train/javascript

number_fn = n => operation => operation ? operation(n) : n;
operator_fn = (callback) => b => a => callback(a, b);

const zero = number_fn(0);
const one = number_fn(1);
const two = number_fn(2);
const three = number_fn(3);
const four = number_fn(4);
const five = number_fn(5);
const six = number_fn(6);
const seven = number_fn(7);
const eight = number_fn(8);
const nine = number_fn(9);

const plus = operator_fn((a, b) => a + b);
const minus = operator_fn((a, b) => a - b);
const times = operator_fn((a, b) => a * b);
const dividedBy = operator_fn((a, b) => Math.floor(a / b));

// ----------------------------------------------------------------------------

const test = (assertion, expected) => ({ assertion, expected });

const TESTS = [
  test(seven(times(five())), 35),
  test(zero(minus(eight())), -8),
  test(four(plus(nine())), 13),
  test(eight(minus(three())), 5),
  test(six(dividedBy(two())), 3),
  test(six(dividedBy(nine())), 0),
  test(seven(dividedBy(four())), 1),
];

for (const test in TESTS) {
  const { assertion, expected } = test;
  console.log(assertion === expected ? 'PASSED' : 'NOT PASSED');
}

File: codewars.com/python/5kyu/calculating_functions.py
# https://www.codewars.com/kata/525f3eda17c7cd9f9e000b39/train/python

zero = lambda operation = None: operation(0) if operation is not None else 0
one = lambda operation = None: operation(1) if operation is not None else 1
two = lambda operation = None: operation(2) if operation is not None else 2
three = lambda operation = None: operation(3) if operation is not None else 3
four = lambda operation = None: operation(4) if operation is not None else 4
five = lambda operation = None: operation(5) if operation is not None else 5
six = lambda operation = None: operation(6) if operation is not None else 6
seven = lambda operation = None: operation(7) if operation is not None else 7
eight = lambda operation = None: operation(8) if operation is not None else 8
nine = lambda operation = None: operation(9) if operation is not None else 9

plus = lambda b: lambda a: a + b
minus = lambda b: lambda a: a - b
times = lambda b: lambda a: a * b
divided_by = lambda b: lambda a: a // b

tests = [
    {
        'assertion': seven(times(five())),
        'expected': 35
    },
    {
        'assertion': four(plus(nine())),
        'expected': 13
    },
    {
        'assertion': eight(minus(three())),
        'expected': 5
    },
    {
        'assertion': six(divided_by(two())),
        'expected': 3
    },
]

for test in tests:
    print('PASSED' if test['assertion'] == test['expected'] else 'NOT PASSED')
